{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "hairGAN.ipynb",
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "aIdhnNECWzvS",
        "outputId": "e075ab76-5d14-47a9-9497-819e2e0404b4"
      },
      "source": [
        "import torch\n",
        "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
        "print('Using device:', device)\n",
        "print()\n",
        "\n",
        "#Additional Info when using cuda\n",
        "if device.type == 'cuda':\n",
        "    print(torch.cuda.get_device_name(0))\n",
        "    print('Memory Usage:')\n",
        "    print('Allocated:', round(torch.cuda.memory_allocated(0)/1024**3,1), 'GB')\n",
        "    print('Cached:   ', round(torch.cuda.memory_reserved(0)/1024**3,1), 'GB')"
      ],
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Using device: cuda\n",
            "\n",
            "Tesla T4\n",
            "Memory Usage:\n",
            "Allocated: 0.0 GB\n",
            "Cached:    0.0 GB\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "p--cGeLKa7Kl",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "295da706-761f-4036-dc88-67863d6ba079"
      },
      "source": [
        "!apt install git-lfs"
      ],
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following NEW packages will be installed:\n",
            "  git-lfs\n",
            "0 upgraded, 1 newly installed, 0 to remove and 40 not upgraded.\n",
            "Need to get 2,129 kB of archives.\n",
            "After this operation, 7,662 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 git-lfs amd64 2.3.4-1 [2,129 kB]\n",
            "Fetched 2,129 kB in 1s (1,669 kB/s)\n",
            "Selecting previously unselected package git-lfs.\n",
            "(Reading database ... 148492 files and directories currently installed.)\n",
            "Preparing to unpack .../git-lfs_2.3.4-1_amd64.deb ...\n",
            "Unpacking git-lfs (2.3.4-1) ...\n",
            "Setting up git-lfs (2.3.4-1) ...\n",
            "Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EYxEbqrla9D-",
        "outputId": "e800adc9-7a22-47e6-e43a-c99ace0d5095"
      },
      "source": [
        "!git lfs clone https://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@github.com/vanquish630/hairGAN.git"
      ],
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "WARNING: 'git lfs clone' is deprecated and will not be updated\n",
            "          with new flags from 'git clone'\n",
            "\n",
            "'git clone' has been updated in upstream Git to have comparable\n",
            "speeds to 'git lfs clone'.\n",
            "Cloning into 'hairGAN'...\n",
            "remote: Enumerating objects: 103, done.\u001b[K\n",
            "remote: Counting objects: 100% (103/103), done.\u001b[K\n",
            "remote: Compressing objects: 100% (87/87), done.\u001b[K\n",
            "remote: Total 103 (delta 12), reused 101 (delta 10), pack-reused 0\u001b[K\n",
            "Receiving objects: 100% (103/103), 731.49 KiB | 7.62 MiB/s, done.\n",
            "Resolving deltas: 100% (12/12), done.\n",
            "Git LFS: (7 of 7 files) 1.07 GB / 1.07 GB\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "BBdrxq4_bEzZ",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "2c0349cf-2866-4ef3-d9f1-93e932684c8d"
      },
      "source": [
        "!pip install face_alignment\n"
      ],
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: face_alignment in /usr/local/lib/python3.7/dist-packages (1.3.4)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from face_alignment) (1.19.5)\n",
            "Requirement already satisfied: opencv-python in /usr/local/lib/python3.7/dist-packages (from face_alignment) (4.1.2.30)\n",
            "Requirement already satisfied: torch in /usr/local/lib/python3.7/dist-packages (from face_alignment) (1.9.0+cu102)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.7/dist-packages (from face_alignment) (4.62.0)\n",
            "Requirement already satisfied: scipy>=0.17 in /usr/local/lib/python3.7/dist-packages (from face_alignment) (1.4.1)\n",
            "Requirement already satisfied: numba in /usr/local/lib/python3.7/dist-packages (from face_alignment) (0.51.2)\n",
            "Requirement already satisfied: scikit-image in /usr/local/lib/python3.7/dist-packages (from face_alignment) (0.16.2)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.7/dist-packages (from numba->face_alignment) (57.4.0)\n",
            "Requirement already satisfied: llvmlite<0.35,>=0.34.0.dev0 in /usr/local/lib/python3.7/dist-packages (from numba->face_alignment) (0.34.0)\n",
            "Requirement already satisfied: pillow>=4.3.0 in /usr/local/lib/python3.7/dist-packages (from scikit-image->face_alignment) (7.1.2)\n",
            "Requirement already satisfied: matplotlib!=3.0.0,>=2.0.0 in /usr/local/lib/python3.7/dist-packages (from scikit-image->face_alignment) (3.2.2)\n",
            "Requirement already satisfied: PyWavelets>=0.4.0 in /usr/local/lib/python3.7/dist-packages (from scikit-image->face_alignment) (1.1.1)\n",
            "Requirement already satisfied: imageio>=2.3.0 in /usr/local/lib/python3.7/dist-packages (from scikit-image->face_alignment) (2.4.1)\n",
            "Requirement already satisfied: networkx>=2.0 in /usr/local/lib/python3.7/dist-packages (from scikit-image->face_alignment) (2.6.2)\n",
            "Requirement already satisfied: python-dateutil>=2.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib!=3.0.0,>=2.0.0->scikit-image->face_alignment) (2.8.2)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.7/dist-packages (from matplotlib!=3.0.0,>=2.0.0->scikit-image->face_alignment) (0.10.0)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib!=3.0.0,>=2.0.0->scikit-image->face_alignment) (1.3.1)\n",
            "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib!=3.0.0,>=2.0.0->scikit-image->face_alignment) (2.4.7)\n",
            "Requirement already satisfied: six in /usr/local/lib/python3.7/dist-packages (from cycler>=0.10->matplotlib!=3.0.0,>=2.0.0->scikit-image->face_alignment) (1.15.0)\n",
            "Requirement already satisfied: typing-extensions in /usr/local/lib/python3.7/dist-packages (from torch->face_alignment) (3.7.4.3)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4S0NBUm6hRC0",
        "outputId": "55fcef8c-c2b7-4850-d6a8-2d0cd90a7940"
      },
      "source": [
        "%cd hairGAN/"
      ],
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/hairGAN/hairGAN\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "sRqqxIABkzTa"
      },
      "source": [
        "## Hair removal"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xua4gSF3ksII",
        "outputId": "23d2726c-5610-4ad5-9ec7-eeebd15b32e3"
      },
      "source": [
        "!python test.py --test_dir test_data"
      ],
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[2021-09-04 09:43:54,324][INFO] logging into ./results/hairGAN.log\n",
            "[2021-09-04 09:43:54,340][INFO] found 7 images in the folder\n",
            "[2021-09-04 09:43:54,340][INFO] Loading models for stylegan inversion.\n",
            "[2021-09-04 09:43:54,340][INFO] Loading inverter model.\n",
            "[2021-09-04 09:43:54,340][INFO] Build network for module `generator` in model `styleganinv_ffhq256`.\n",
            "[2021-09-04 09:43:54,563][INFO] Loading pytorch weights from `/content/hairGAN/hairGAN/pretrained_models/styleganinv_ffhq256_generator.pth`.\n",
            "[2021-09-04 09:44:01,402][INFO] Successfully loaded!\n",
            "[2021-09-04 09:44:01,406][INFO] Current `lod` is 0.0.\n",
            "[2021-09-04 09:44:01,406][INFO] Build network for module `encoder` in model `styleganinv_ffhq256`.\n",
            "[2021-09-04 09:44:02,601][INFO] Loading pytorch weights from `/content/hairGAN/hairGAN/pretrained_models/styleganinv_ffhq256_encoder.pth`.\n",
            "[2021-09-04 09:44:03,447][INFO] Successfully loaded!\n",
            "[2021-09-04 09:44:04,378][INFO] Loading generator model.\n",
            "[2021-09-04 09:44:04,378][INFO] Build network for module `generator` in model `styleganinv_ffhq256`.\n",
            "[2021-09-04 09:44:04,568][INFO] Loading pytorch weights from `/content/hairGAN/hairGAN/pretrained_models/styleganinv_ffhq256_generator.pth`.\n",
            "[2021-09-04 09:44:04,732][INFO] Successfully loaded!\n",
            "[2021-09-04 09:44:04,734][INFO] Current `lod` is 0.0.\n",
            "[2021-09-04 09:44:04,734][INFO] Loading models for hypergan inpainting.\n",
            "2021-09-04 09:44:04.794736: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-09-04 09:44:04.948394: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-09-04 09:44:04.948974: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-09-04 09:44:04.949914: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-09-04 09:44:04.950454: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-09-04 09:44:04.950931: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-09-04 09:44:07.463973: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-09-04 09:44:07.464672: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-09-04 09:44:07.465408: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-09-04 09:44:07.465892: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:39] Overriding allow_growth setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.\n",
            "2021-09-04 09:44:07.467046: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1510] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 11103 MB memory:  -> device: 0, name: Tesla T4, pci bus id: 0000:00:04.0, compute capability: 7.5\n",
            "2021-09-04 09:44:09.540890: W tensorflow/core/framework/cpu_allocator_impl.cc:80] Allocation of 18874368 exceeds 10% of free system memory.\n",
            "2021-09-04 09:44:09.751728: W tensorflow/core/framework/cpu_allocator_impl.cc:80] Allocation of 18874368 exceeds 10% of free system memory.\n",
            "Downloading: \"https://download.pytorch.org/models/resnet18-5c106cde.pth\" to /root/.cache/torch/hub/checkpoints/resnet18-5c106cde.pth\n",
            "100% 44.7M/44.7M [00:00<00:00, 199MB/s]\n",
            "/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py:718: UserWarning: Named tensors and all their associated APIs are an experimental feature and subject to change. Please do not use them for anything important until they are released as stable. (Triggered internally at  /pytorch/c10/core/TensorImpl.h:1156.)\n",
            "  return torch.max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode)\n",
            "2021-09-04 09:44:13.737792: I tensorflow/stream_executor/cuda/cuda_dnn.cc:369] Loaded cuDNN version 8004\n",
            "[2021-09-04 09:44:30,801][INFO] succesfully completed inpainting on 7 / 7 images\n",
            "[2021-09-04 09:44:30,802][INFO] starting inversion!\n",
            "Building inverter\n",
            "loss_pix: 0.008, loss_ssim: 0.809, loss_feat: 1344.253, loss_reg: 0.037, loss: -2.330: 100% 200/200 [00:36<00:00,  5.45it/s]\n",
            "loss_pix: 0.143, loss_ssim: 0.524, loss_feat: 3605.576, loss_reg: 22.196, loss: 12.069: 100% 200/200 [00:37<00:00,  5.36it/s]\n",
            "loss_pix: 0.002, loss_ssim: 0.877, loss_feat: 771.173, loss_reg: 0.027, loss: -2.574: 100% 200/200 [00:37<00:00,  5.28it/s]\n",
            "loss_pix: 0.003, loss_ssim: 0.837, loss_feat: 1131.568, loss_reg: 0.040, loss: -2.426: 100% 200/200 [00:38<00:00,  5.21it/s]\n",
            "loss_pix: 0.019, loss_ssim: 0.675, loss_feat: 2137.795, loss_reg: 0.051, loss: -1.868: 100% 200/200 [00:38<00:00,  5.15it/s]\n",
            "loss_pix: 0.215, loss_ssim: 0.239, loss_feat: 6341.285, loss_reg: 0.577, loss: 0.160: 100% 200/200 [00:39<00:00,  5.13it/s]\n",
            "loss_pix: 0.003, loss_ssim: 0.814, loss_feat: 1083.711, loss_reg: 0.035, loss: -2.365: 100% 200/200 [00:38<00:00,  5.14it/s]\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-0\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-1\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-2\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-0\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-4\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-5\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-6\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-7\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-1\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-9\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-10\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-11\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-12\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-2\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-14\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-15\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-16\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-17\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-3\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-19\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-20\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-21\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-22\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-4\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-24\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-25\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-26\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-27\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-5\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-29\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-30\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-31\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-32\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-6\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-34\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-35\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-36\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-37\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-0._init_input_shape\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer-1._init_input_shape\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-0.kernel\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-0.bias\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-1.kernel\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-1.bias\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-2.kernel\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-2.bias\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-3.kernel\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-3.bias\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-4.kernel\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-4.bias\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-5.kernel\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-5.bias\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-6.kernel\n",
            "WARNING:tensorflow:Unresolved object in checkpoint: (root).discriminator.layer_with_weights-6.bias\n",
            "WARNING:tensorflow:A checkpoint was restored (e.g. tf.train.Checkpoint.restore or tf.keras.Model.load_weights) but not all checkpointed values were used. See above for specific issues. Use expect_partial() on the load status object, e.g. tf.train.Checkpoint.restore(...).expect_partial(), to silence these warnings, or use assert_consumed() to make the check explicit. See https://www.tensorflow.org/guide/checkpoint#loading_mechanics for details.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "sGvKzRtVk5FI"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}